### ISSUE TITLE
How to Change 'gatsby build' directory？

### ISSUE THREAD

**Issue by PiccoloYu** [gatsbyjs/gatsby#18975]
How to Change 'gatsby build' directory？

**Comment by lannonbr** [comment]
So when `gatsby build` runs, it will output all the files you need to deploy to the `public` folder inside the root of the site folder. That said, after the site is built, you could just take the contents of such and move it into another directory completely.

**Comment by sidharthachatterjee** [comment]
@PiccoloYu We currently don't support a configurable output directory and don't plan to doing so in the near future. 

Let's follow along the conversation in https://github.com/gatsbyjs/gatsby/issues/1878

Closing this

**Comment by abraaoz** [comment]
Yes, you can.
To change the folder from `public` to `dist`, add the following code in your `gatsby-node.js`:
```javascript
const path = require("path")
const fs = require("fs")

exports.onPreInit = () => {
  if (process.argv[2] === "build") {
    fs.rmdirSync(path.join(__dirname, "dist"), { recursive: true })
    fs.renameSync(
      path.join(__dirname, "public"),
      path.join(__dirname, "public_dev")
    )
  }
}

exports.onPostBuild = () => {
  fs.renameSync(path.join(__dirname, "public"), path.join(__dirname, "dist"))
  fs.renameSync(
    path.join(__dirname, "public_dev"),
    path.join(__dirname, "public")
  )
}
```

**Comment by 211211** [comment]
Your solution works for me @abraaoz.
Thank you very much.

**Comment by seyacat** [comment]
based on abraoz y resolve like this

const path = require("path")
const fs = require("fs-extra")


exports.onPostBuild = () => {
  fs.moveSync(path.join(__dirname, "public"), path.join(__dirname, "../public"),{ overwrite: true })

}

**Comment by noorvir** [comment]
@abraaoz's solution works great. However, `gatsby serve` now fails because it tries to serve the build from the `/public` directory. 

Is there a way to override this as well? The [docs](https://www.gatsbyjs.org/docs/gatsby-cli/#serve) don't seem to provide an option to serve a specific path.

**Comment by martin-braun** [comment]
If you need the files in a different directory after build, you can

```sh
npm i rsync -D
```

and modify the `npm run build` script to do

```sh
gatsby build && rsync -av --delete public/ path/to/new/folder
```

This will make sure that _path/to/new/folder_ does not contain files, that are not in _public/_ and copies files over in one command. Using the _rsync_ npm package will ensure cross-compatibility.

**Comment by jamo** [comment]
Hey all, could you help me understand what would be the use-case for making the build path configurable?

**Comment by eugenedm** [comment]
> Hey all, could you help me understand what would be the use-case for making the build path configurable?

I think this WOULD help a lot... I run gatsby for my blog on "/blog" and I run rails app on my "/" root... So, this is a need if you have multiple applications running on the same domain... I can't use a subdomain because I want to rank on SEO...

**Comment by y1n0** [comment]
I believe the `gatsby build` command should have an option to configure this

**Comment by jolares** [comment]
> Hey all, could you help me understand what would be the use-case for making the build path configurable?

for sure, projects with more complex build/scaffolding/deployment/serving requirements, such as those in many enterprise projects, could find this useful. For example, an organization using monorepo design patterns and having a standard build process (e.g distribution bundles being built to some other common location in the repository before getting picked up, optimized, combined, or served by some other processes). 

Basically without this capability, orgs need to then customize the build/serve/deploy tool just to handle this path instead (which is not a big deal btw, but i am just providing use-cases).

**Comment by willkuerlich** [comment]
My use-case would be, to be able to keep my Gatsby develop process running (which is pretty time consuming to start due to complex sourcing etc) while being able to create build while developing. Otherwise I need to do this by using the same git branch in 2 directories (1 directory for using for my actual local development and 1 directory for building the development environment online).
More of a development comfort issue here, but nevertheless ;)

**Comment by biel-fyq** [comment]
![image](https://user-images.githubusercontent.com/86816096/194810853-bf96b3ed-c028-46e0-ad15-351f96d98cea.png)
![image](https://user-images.githubusercontent.com/86816096/194811662-936f238a-48f6-4b05-af99-cd3a1874d56a.png)
why？

**Comment by martin-braun** [comment]
@biel-fyq How about translating the error, so people can help?

**Comment by ra9r** [comment]
I would like to see this feature as well.  In my situation I use gatsby to generate the website in a Firebase hosting project, but the project that contains the website also contains the code for functions, security rules, storage configuration and a host of other directories.  Having Gatsby enforce an out directory that is not configurable makes it necessary to move/delete the public directory to some where that I need it.  It's a very common thing to do in anything except the most basic websites.  You really should consider adding this feature.

**Comment by technoknol** [comment]
> Yes, you can. To change the folder from `public` to `dist`, add the following code in your `gatsby-node.js`:
> 
> ```js
> const path = require("path")
> const fs = require("fs")
> 
> exports.onPreInit = () => {
>   if (process.argv[2] === "build") {
>     fs.rmdirSync(path.join(__dirname, "dist"), { recursive: true })
>     fs.renameSync(
>       path.join(__dirname, "public"),
>       path.join(__dirname, "public_dev")
>     )
>   }
> }
> 
> exports.onPostBuild = () => {
>   fs.renameSync(path.join(__dirname, "public"), path.join(__dirname, "dist"))
>   fs.renameSync(
>     path.join(__dirname, "public_dev"),
>     path.join(__dirname, "public")
>   )
> }
> ```

here's the new version that won't throw an error: 

```javascript
const fs = require('fs');
const path = require('path');

exports.onPreInit = () => {
  if (process.argv[2] === "build") {
    try {
      if (fs.existsSync(path.join(__dirname, "dist"))) { // Check if dist exists before attempting to remove
        fs.rmdirSync(path.join(__dirname, "dist"), { recursive: true });
        console.log("Removed existing dist directory.");
      } else {
        console.log("No existing dist directory to remove.");
      }
    } catch (err) {
      console.error("Error removing dist directory:", err);
      // Handle the error appropriately. For example, you might want to exit the process:
      process.exit(1); // Exit with error code
    }

    try {
      if (fs.existsSync(path.join(__dirname, "public"))) { // Check if public exists before renaming
        fs.renameSync(
          path.join(__dirname, "public"),
          path.join(__dirname, "public_dev")
        );
        console.log("Renamed public to public_dev.");
      } else {
        console.error("public directory not found. Cannot rename.");
        process.exit(1);
      }
    } catch (err) {
      console.error("Error renaming public directory:", err);
      process.exit(1);
    }
  }
};

exports.onPostBuild = () => {
    try {
        if(fs.existsSync(path.join(__dirname, "public"))){
            fs.renameSync(path.join(__dirname, "public"), path.join(__dirname, "dist"));
            console.log("Renamed public to dist.");
        }else{
            console.error("public directory not found. Cannot rename.");
            process.exit(1);
        }
    } catch (err) {
        console.error("Error renaming public to dist:", err);
        process.exit(1);
    }

    try {
        if(fs.existsSync(path.join(__dirname, "public_dev"))){
            fs.renameSync(
                path.join(__dirname, "public_dev"),
                path.join(__dirname, "public")
            );
            console.log("Renamed public_dev to public.");
        }else{
            console.error("public_dev directory not found. Cannot rename.");
            process.exit(1);
        }
    } catch (err) {
        console.error("Error renaming public_dev to public:", err);
        process.exit(1);
    }
};
```
